mediacloud/mediawords/dbi/downloads.py

```diff
@@ -0,0 +1,388 @@
+
+
+def fetch_content(db: DatabaseHandler, download: dict) -> str:
+    """Fetch the content for the given download from the configured content store."""
+    if 'downloads_id' not in download:
+        raise McDBIDownloadsException("downloads_id not in download")
+
+    if not download_successful(download):
+        raise McDBIDownloadsException(
+            "attempt to fetch content for unsuccessful download: %d" % (download['downloads_id']))
+
+    store = _get_store_for_reading(download)
+
+    content_bytes = store.fetch_content(db, download['downloads_id'], download['path'])
+
+    content = content_bytes.decode()
+
+    # horrible hack to fix old content that is not stored in unicode
+    config = mediawords.util.config.get_config()
+    ascii_hack_downloads_id = config['mediawords'].get('ascii_hack_downloads_id', 0)
+    if download['downloads_id'] < ascii_hack_downloads_id:
+        # this matches all non-printable-ascii characters.  python re does not support POSIX character
+        # classes like [[:ascii:]]
+        content = re.sub(r'[^ -~]', ' ', content)
+
+    return content
+
+
+def store_content(db: DatabaseHandler, download: dict, content: str) -> dict:
+    """Store the content for the download."""
+    # feed_error state indicates that the download was successfull but that there was a problem
+    # parsing the feed afterward.  so we want to keep the feed_error state even if we redownload
+    # the content
+    new_state = 'success' if download['state'] != 'feed_error' else 'feed_error'
+
+    try:
+        path = _get_store_for_writing().store_content(db, download['downloads_id'], content)
+        error = ''
+    except Exception as e:
+        raise McDBIDownloadsException("error while trying to store download %d: %s" % (download['downloads_id'], e))
+        new_state = 'error'
+        error = str(e)
+        path = ''
+
+    if new_state == 'success':
+        error = ''
+
+    db.update_by_id('downloads', download['downloads_id'], {'state': new_state, 'path': path, 'error_message': error})
+
+    download = db.find_by_id('downloads', download['downloads_id'])
+
+    return download
+
+
+def _get_cached_extractor_results(db: DatabaseHandler, download: dict) -> typing.Optional[dict]:
+    """Get extractor results from cache.
+
+    Return:
+    None if there is a miss or a dict in the form of extract_content() if there is a hit.
+    """
+    r = db.query(
+        "select extracted_html, extracted_text from cached_extractor_results where downloads_id = %(a)s",
+        {'a': download['downloads_id']}).hash()
+
+    log.debug("EXTRACTOR CACHE HIT" if r is not None else "EXTRACTOR CACHE MISS")
+
+    return r
+
+
+def _set_cached_extractor_results(db, download, results) -> None:
+    """Store results in extractor cache and manage size of cache."""
+    max_cache_entries = 1000 * 1000
+
+    # occasionally delete too old entries in the cache
+    if random.random() * (max_cache_entries / 10) < 1:
```

Comment:
I have added more comments to the code to explain what's going on here.  We only need on the order of ten thousand rows in this table to handle its job as a sort of working memory of extractions within the topic spider, but it is pretty trivial for us to host a million-ish row table on our postgres install.  It is expensive (in query time or denormalization  complexity) to directly keep track of a postgres table size, though, so it is simplest and quickest just to trim the table every 100k requests. 

We could use some sort of cache module, but we need this to be in some common data store shared between servers, and postgres is our general purpose robust solution for a common data store.  I could use some sort of caching module that hides this logic, but this is simple, robust, and transparent in a critical code path.

I disagree that it is unintuitive that a set_cache_value() function will be responsible for trimming old data from the cache.  The only alternatives are either requiring the user to periodically run a 'trim_cache()' function or to have a daemon running somewhere to manage the cache, both of which are more complex and fragile.

Reply:
We already have a PostgreSQL-based object cache that's being purged by a separate `purge_object_caches.py` script run by Supervisor, so it would be nice to reuse that cache here somehow. But again, to be able to do that, we want this and other code to be in Python first!